### PR TITLE
[new release] atacama (0.0.5)

### PR DESCRIPTION
**Pull request by leostera**
Modern, pure OCaml socket pool for Riot

- Project page: <a href="https://github.com/suri-framework/atacama">https://github.com/suri-framework/atacama</a>

##### CHANGES:

* Bring up to latest Riot version – thanks @metame! :clap:

* Introduce `handle_message` callback to allow connector processes to receive
  out-of-band messages from other Riot processes

* Introduce configurable transports with their own configruation

* Introduce configurable connection limits and a dynamic connection pool

* Introduce SSL Transport

* Introduce Connection Peer data including the address and port

* Expose new connection measurements such as time the connection was accepted
  (`accepted_at`) and time at which the connection was established (`connected_at`)

* Move to bytestrings and iovecs

* Default to not reusing ports

### CODE REVIEW

**Comment by leostera** [comment]
Depends on #25654

**Comment by raphael-proust** [comment]
There are CI failures for all the platforms:

```
#=== ERROR while compiling atacama.0.0.5 ======================================#
# context              2.2.0~beta3~dev | linux/x86_64 | ocaml-options-only-afl.1 | pinned(https://github.com/suri-framework/atacama/releases/download/0.0.5/atacama-0.0.5.tbz)
# path                 ~/.opam/5.1/.opam-switch/build/atacama.0.0.5
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p atacama -j 255 @install
# exit-code            1
# env-file             ~/.opam/log/atacama-7-7b8a9e.env
# output-file          ~/.opam/log/atacama-7-7b8a9e.out
### output ###
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlc.opt -w -40 -g -bin-annot -I atacama/.atacama.objs/byte -I /home/opam/.opam/5.1/lib/angstrom -I /home/opam/.opam/5.1/lib/bigstringaf -I /home/opam/.opam/5.1/lib/iomux -I /home/opam/.opam/5.1/lib/ocaml/unix -I /home/opam/.opam/5.1/lib/ptime -I /home/opam/.opam/5.1/lib/ptime/clock/os -I /home/opam/.opam/5.1/lib/riot -I /home/opam/.opam/5.1/lib/riot/core -I /home/opam/.opam/5.1/lib/riot/lib -I /home/opam/.opam/5.1/lib/riot/lib/logger -I /home/opam/.opam/5.1/lib/riot/lib/net -I /home/opam/.opam/5.1/lib/riot/log -I /home/opam/.opam/5.1/lib/riot/net -I /home/opam/.opam/5.1/lib/riot/runtime -I /home/opam/.opam/5.1/lib/riot/runtime/time -I /home/opam/.opam/5.1/lib/riot/scheduler -I /home/opam/.opam/5.1/lib/riot/util -I /home/opam/.opam/5.1/lib/stringext -I /home/opam/.opam/5.1/lib/telemetry -I /home/opam/.opam/5.1/lib/uri -no-alias-deps -open Atacama__ -o atacama/.atacama.objs/byte/atacama.cmi -c -intf atacama/atacama.mli)
# File "atacama/atacama.mli", line 6, characters 18-30:
# 6 |   val send : t -> Bytestring.t -> (unit, [> `Closed ]) IO.io_result
#                       ^^^^^^^^^^^^
# Error: Unbound module Bytestring
# (cd _build/default && /home/opam/.opam/5.1/bin/ocamlc.opt -w -40 -g -bin-annot -I atacama/.atacama.objs/byte -I /home/opam/.opam/5.1/lib/angstrom -I /home/opam/.opam/5.1/lib/bigstringaf -I /home/opam/.opam/5.1/lib/iomux -I /home/opam/.opam/5.1/lib/ocaml/unix -I /home/opam/.opam/5.1/lib/ptime -I /home/opam/.opam/5.1/lib/ptime/clock/os -I /home/opam/.opam/5.1/lib/riot -I /home/opam/.opam/5.1/lib/riot/core -I /home/opam/.opam/5.1/lib/riot/lib -I /home/opam/.opam/5.1/lib/riot/lib/logger -I /home/opam/.opam/5.1/lib/riot/lib/net -I /home/opam/.opam/5.1/lib/riot/log -I /home/opam/.opam/5.1/lib/riot/net -I /home/opam/.opam/5.1/lib/riot/runtime -I /home/opam/.opam/5.1/lib/riot/runtime/time -I /home/opam/.opam/5.1/lib/riot/scheduler -I /home/opam/.opam/5.1/lib/riot/util -I /home/opam/.opam/5.1/lib/stringext -I /home/opam/.opam/5.1/lib/telemetry -I /home/opam/.opam/5.1/lib/uri -no-alias-deps -open Atacama__ -o atacama/.atacama.objs/byte/atacama__Connection.cmo -c -impl atacama/connection.ml)
# File "atacama/connection.ml", line 12, characters 20-31:
# 12 |       writer : 'dst IO.Writer.t;
#                          ^^^^^^^^^^^
# Error: Unbound module IO
```

Maybe you forgot to declare a dependency?

**Comment by leostera** [comment]
This depended on Riot 0.0.9, so I'm rerunning the jobs now that that's released! 🙏🏼 Thanks for the ping @raphael-proust

**Comment by samoht** [comment]
So this needs an update in the opam file to specify the lower bound on the latest riot

File: packages/atacama/atacama.0.0.5/opam [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+license: "MIT"
+tags: ["socket" "socket pool" "tcp" "riot"]
+homepage: "https://github.com/suri-framework/atacama"
+bug-reports: "https://github.com/suri-framework/atacama/issues"
+depends: [
+  "dune" {>= "3.12"}
+  "mdx" {with-test & >= "2.3.1"}
+  "ocaml" {>= "5.1"}
+  "odoc" {with-doc & >= "2.2.2"}
+  "riot" {>= "0.0.1"}
```
Comment (samoht) [inline packages/atacama/atacama.0.0.5/opam:16]:
```suggestion
  "riot" {>= "0.0.9"}
```

**Comment by raphael-proust** [comment]
> It seems to be picking up riot 0.0.5 instead of 0.0.9 and I can't find in the logs _why_. @raphael-proust @samoht the bound is `(riot (>= "0.0.1"))` so I'd expect it to pick the latest version that matches this (so 0.0.9 as of yesterday).

Not sure why it's not picking the last one…

> 
> I understand that I could bump the minimum version and use `(riot (>= "0.0.9"))`, but I'd rather not worry about these boundaries and just always pick the latest version.

The issue with this approach is that someone might have riot.0.0.1 installed already, and then they install atacama.0.0.5 and then the install doesn't work. The opam file for atacama should specify which versions of riot are compatible and that means bumping the bound in the file.

When you have multiple packages which are related to the point that they have identical release cycles, then you can use `(= version)` as a special keyword indicating you want the same version number of that package you depend on. But this doesn't seem to be the case here. You seem to want something slightly different but I'm not sure how we could encode "the latest version of that package at the time of release of this package" in a dependency constraint… although maybe you could have some tooling to update your dependency constraints automatically to adhere to that?

**Pull request merged**